File: 42-trapping-rain-water/trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let left = 0
    let right = height.length - 1
    let leftMax = 0, rightMax = 0, water = 0

    while (left < right) {
        if (height[left] <= height[right]) {
            if (height[left] > leftMax) {
                leftMax = height[left]
            } else {
                water += leftMax - height[left]
            }
            left++
        } else {
            if (height[right] > rightMax) {
                rightMax = height[right]
            } else {
                water += rightMax - height[right]
            }
            right--
        }
    }
    return water
};

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0

        count = 0
        left, right = 0, len(height) - 1
        left_max, right_max = 0, 0
        water= 0

        while left <= right:
            left_max = max(left_max, height[left])
            right_max = max(right_max, height[right])

            if left_max <= right_max:
                water += left_max - height[left]
                left += 1
            else:
                water += right_max - height[right]
                right -= 1

        return water